compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -1355,41 +1383,56 @@ impl DepNodeColorMap {
         prev_index: SerializedDepNodeIndex,
         index: DepNodeIndex,
     ) -> Result<(), DepNodeIndex> {
-        let value = &self.values[prev_index];
-        match value.compare_exchange(
-            COMPRESSED_UNKNOWN,
-            index.as_u32(),
-            Ordering::Relaxed,
-            Ordering::Relaxed,
-        ) {
+        use Ordering::*;
+
+        debug_assert!(index != DepNodeIndex::FOREVER_RED_NODE);
+
+        let index = index.as_u32();
+        assert_eq!(index & COMPRESSED_RED_BIT, 0);
+        let green = index;
+
+        match self.values[prev_index].compare_exchange(COMPRESSED_UNKNOWN, green, Release, Acquire)
+        {
             Ok(_) => Ok(()),
-            Err(v) => Err({
-                assert_ne!(v, COMPRESSED_RED, "tried to mark a red node as green");
-                DepNodeIndex::from_u32(v)
-            }),
+            Err(red) if red & COMPRESSED_RED_BIT != 0 => {
+                let green = red & !COMPRESSED_RED_BIT;
+                if let Err(other) =
+                    self.values[prev_index].compare_exchange(red, green, Release, Acquire)
+                {
+                    assert_eq!(other, green)
+                }
+
+                Err(DepNodeIndex::from_u32(green))
+            }
+            Err(green) => Err(DepNodeIndex::from_u32(green)),
```

Comment:
Need comments explaining when all of the cases above may happen, and what we are asserting.